src/dsa/rpo_falcon512/keys/secret_key.rs

```diff
@@ -0,0 +1,400 @@
+        signature::SignaturePoly,
+        ByteReader, ByteWriter, Deserializable, DeserializationError, Nonce, Serializable,
+        ShortLatticeBasis, Signature, Word, MODULUS, N, SIGMA, SIG_L2_BOUND,
+    },
+    PubKeyPoly, PublicKey,
+};
+use crate::dsa::rpo_falcon512::{
+    hash_to_point::hash_to_point_rpo256, math::ntru_gen, SIG_NONCE_LEN, SK_LEN,
+};
+use alloc::{string::ToString, vec::Vec};
+use num::{Complex, Zero};
+use num_complex::Complex64;
+use rand::Rng;
+
+#[cfg(feature = "std")]
+use rand::{rngs::OsRng, RngCore};
+
+#[cfg(not(feature = "std"))]
+use num::Float;
+
+// CONSTANTS
+// ================================================================================================
+
+const WIDTH_BIG_POLY_COEFFICIENT: usize = 8;
+const WIDTH_SMALL_POLY_COEFFICIENT: usize = 6;
+
+// SECRET KEY
+// ================================================================================================
+
+/// The secret key is a quadruple [[g, -f], [G, -F]] of polynomials with integer coefficients. Each
+/// polynomial is of degree at most N = 512 and computations with these polynomials is done modulo
+/// the monic irreducible polynomial ϕ = x^N + 1. The secret key is a basis for a lattice and has
+/// the property of being short with respect to a certain norm and an upper bound appropriate for
+/// a given security parameter. The public key on the other hand is another basis for the same
+/// lattice and can be described by a single polynomial h with integer coefficients modulo ϕ.
+/// The two keys are related by the following relation:
+///
+/// 1. h = g /f [mod ϕ][mod p]
+/// 2. f.G - g.F = p [mod ϕ]
+///
+/// where p = 12289 is the Falcon prime. Equation 2 is called the NTRU equation.
+/// The secret key is generated by first sampling a random pair (f, g) of polynomials using
+/// an appropriate distribution that yields short but not too short polynomials with integer
+/// coefficients modulo ϕ. The NTRU equation is then used to find a matching pair (F, G).
+/// The public key is then derived from the secret key using equation 2.
+///
+/// To allow for fast signature generation, the secret key is pre-processed into a more suitable
+/// form, called the LDL tree, and this allows for fast sampling of short vectors in the lattice
+/// using Fast Fourier sampling during signature generation (ffSampling algorithm 11 in [1]).
+///
+/// [1]: https://falcon-sign.info/falcon.pdf
+#[derive(Debug, Clone)]
+pub struct SecretKey {
+    secret_key: ShortLatticeBasis,
+    tree: LdlTree,
+}
+
+#[allow(clippy::new_without_default)]
+impl SecretKey {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Generates a secret key from OS-provided randomness.
+    #[cfg(feature = "std")]
+    pub fn new() -> Self {
+        let mut seed: [u8; 32] = [0; 32];
+        OsRng.fill_bytes(&mut seed);
+        Self::with_rng(&mut OsRng)
+    }
```

Comment:
I put this constructor under `std` flag. In `no_std` context, we could use `with_rng()` constructor instead.